Арсений Карпов

jexinox opened the pull request:
@neoezop

jexinox committed:
Added TagsCloudVisualization.csproj
Added CircularCloudLayouter.cs
Added CircularCloudLayouter_Should.cs

jexinox committed:
Added ILayouter.cs

jexinox committed:
Added SpiralLayouter.cs
Changed CircularCloudLayouter.cs

jexinox committed:
Renamed ILayouter.cs to IRectangleLayouter.cs
Added Spiral and PointD classes

jexinox committed:
Added Rectangle Extensions to check intersections
Added Spiral IncreaseSize method and constructor

jexinox committed:
Sorted files to dirs
Added TestFixtures

jexinox committed:
Added SpiralRectangleLayouter TestFixture

jexinox committed:
Moved Spiral.cs SpiralRectangleLayouter.cs IRectangleLayouter.cs to RectangleLayouters folder

jexinox committed:
Added Spiral Tests

neoezop commented on cs/TagsCloudVisualization/Extensions/RectangleExtensions.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    internal static class RectangleExtensions
+    {
+        public static bool Intersects(this Rectangle firstRect, Rectangle secondRect)
```
У Rectangle есть свой IntersectsWith, он не подойдёт?

jexinox replied:
Да, он подойдет, не знал о нем, всегда расширение писал

neoezop commented on cs/TagsCloudVisualization/Helpers/PointD.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace TagsCloudVisualization.Helpers
+{
+    internal class PointD
```
Почему решил сделать свой класс, а не заиспользовать точку из System.Windows?

jexinox replied:
Подумал, что та точка используется только в WPF и почему-то решил не использовать, когда задали вопрос, стало ясно, что это так себе аргумент и какой смысл создавать свою, если есть уже готовая во встроенной библиотеке...

jexinox replied:
Когда открыл проект вспомнил зачем создал собственный PointD, если его не создать, то возникает конфликт между System.Drawing.Point и System.Windows.Point

neoezop commented on cs/TagsCloudVisualization/Helpers/PointD.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace TagsCloudVisualization.Helpers
+{
+    internal class PointD
```
Кстати, можно переделать в структуру :)

neoezop commented on cs/TagsCloudVisualization/Layouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace TagsCloudVisualization.Layouters
+{
+    public class CircularCloudLayouter
```
Пока не очень понятно зачем нужен ещё один layouter

jexinox replied:
Я думал создать что-то вроде гибкой структуры, то есть в CircularCloudLayouter передавать класс реализующий интерфейс ILayouter, но потом еще раз все обдумал и в этом нет особого смысла, лучше сделать так с классом, который будет рисовать с помощью ILayouter'a

neoezop commented on cs/TagsCloudVisualization/Tests/Spiral_Should.cs:
```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections;
```
Неиспользуемая либа

neoezop commented on cs/TagsCloudVisualization/Tests/Spiral_Should.cs:
```diff
@@ -0,0 +1,89 @@
+namespace TagsCloudVisualization.Tests
+{
+    [TestFixture]
+    public class Spiral_Should
```
В тестах сейчас проверяется только то, что спиралька работает по формуле, которую мы туда и вбили (причём как мы получаем ожидаемые результаты при этом не описано). Это, конечно, не совсем бесполезно, но и не очень полезно. В случае со спиралькой больше хотелось бы видеть, например, что она располагает точки с удовлетворяющей нас плотностью

neoezop commented on cs/TagsCloudVisualization/Tests/Spiral_Should.cs:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        [Test]
+        [TestCaseSource(nameof(CenterGenerator))]
+        public void IncreaseSize_ChangingPoint_WithCustomValues(Point center)
+        {
+            var spiral = new Spiral(center);
+            
+            spiral.IncreaseSize(1, 1);
+            
+            spiral.CurrentPoint.X.Should().BeApproximately(Math.Cos(1) * 1 + center.X, 1e-9);
+            spiral.CurrentPoint.Y.Should().BeApproximately(Math.Sin(1) * 1 + center.Y, 1e-9);
+        }
+
+        [Test]
+        [TestCaseSource(nameof(CenterGenerator))]
+        public void IncreaseSize_ChangingPointRepeatedly_WithDefaultValues(Point center)
+        {
+            var spiral = new Spiral(center);
+            
+            spiral.IncreaseSize();
+            spiral.IncreaseSize();
+            spiral.IncreaseSize();
+            
+            spiral.CurrentPoint.X.Should().BeApproximately(Math.Cos(0.3) * 0.3 + center.X, 1e-9);
+            spiral.CurrentPoint.Y.Should().BeApproximately(Math.Sin(0.3) * 0.3 + center.Y, 1e-9);
+        }
+        
+        [Test]
+        [TestCaseSource(nameof(CenterGenerator))]
+        public void IncreaseSize_ChangingPointRepeatedly_WithCustomValues(Point center)
+        {
+            var spiral = new Spiral(center);
+            
+            spiral.IncreaseSize(1, 2);
+            spiral.IncreaseSize(3, 4);
+            spiral.IncreaseSize(5, 6);
+            
+            spiral.CurrentPoint.X.Should().BeApproximately(Math.Cos(12) * 9 + center.X, 1e-9);
+            spiral.CurrentPoint.Y.Should().BeApproximately(Math.Sin(12) * 9 + center.Y, 1e-9);
```
1e-9 во всех тестах используется, можно вынести в константу с говорящим именем

neoezop replied:
Значения в ShouldBeApproximately сам посчитал? Давай вынесем формулу

neoezop reviewed (commented):
Ещё стоит унести всё это в отдельный solution и тесты выделить в другой проект

jexinox committed:
Moved homework to Homework.sln

jexinox committed:
Moved tests to new project
Refactored namespaces
Deleted RectangleExtensions.cs
Deleted IRectangleLayouter.cs and SpiralRectangleLayouter.cs
Deleted SpiralRectangleLayouter_Should.cs

jexinox committed:
Small fixes

jexinox committed:
Added constants
Added methods for counting expected spiral X and Y

jexinox committed:
Removed PointD.cs
Changed Point and Rectangle to PointF and RectangleF because of big problems with double compatibility with System.Drawing
Changed TagsCloudVisualizationTests.csproj because of problems with tests running (I dont know why but Rider thought that I somehow downgraded NUnit and SDK for tests, maybe because of moving new solution to .net5.0)

jexinox committed:
Added putting rectangles by spiral

jexinox committed:
Fixed infinite cycle in CircularCloudLayouter.cs FindRectanglePosition
Moved IsRectangleIntersectsWithAny from CircularCloudLayouter.cs to RectangleFExtensions.cs

jexinox committed:
Made Spiral internal and visible to tests

jexinox committed:
Fixed CircularCloudLayouter.cs doesn't layouts rectangles on center of spiral points

jexinox committed:
Added GetArea and GetCenterPoint in RectangleFExtensions.cs
Added PointFExtensions.cs with GetDistanceTo
Changed Spiral.cs IncreaseSize default increase values
Added tests for CircularCloudLayouter.cs
Added Generators.cs and moved one from Spiral_Should.cs

jexinox committed:
Added console app for testing
Added PutRectangles method in CircularCloudLayouter
Added CircularCloudVisualizator
Refactored CircularCloudLayoter tests
Small fixes in other files

jexinox committed:
Added tests layouting when test failed
Removed layouter from visualizator
Changed rectangles collection in layouter to public

jexinox committed:
Added examples of layouts

jexinox committed:
Refactor readme.md

jexinox committed:
Короче я не особо понимаю, что с маркдауном, попробовал вот так исправить

jexinox committed:
Теперь работае :thumb

jexinox committed:
Update README.md

neoezop commented on Homework/TagsCloudVisualization/TagsCloudVisualization.csproj:
```diff
@@ -0,0 +1,14 @@
+    <ItemGroup>
+      <PackageReference Include="FluentAssertions" Version="6.2.0" />
+      <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
+      <PackageReference Include="NUnit" Version="3.13.2" />
```
Лишние Package-и

neoezop commented on Homework/ConsoleAppForUsingTests/Program.cs:
```diff
@@ -0,0 +1,42 @@
+using TagsCloudVisualization.Visualization;
+using TagsCloudVisualizationTests;
+
+namespace ConsoleAppForTests
```
Давай его немного переименуем и сделаем из этого проекта простую консольку, чтобы получать картинки (сложная консолька будет в следующих TagCloud-ах)

neoezop commented on Homework/ConsoleAppForUsingTests/Program.cs:
```diff
@@ -0,0 +1,42 @@
+{
+    internal class Program
+    {
+        static void Main(string[] args)
+        {
+            var layouter = new CircularCloudLayouter(new Point(400, 300));
+            var visualizator = new CircularCloudVisualizator();
+
+            var rectangles = new List<RectangleF>();
+
+            for (var i = 0; i < 100; ++i)
+            {
+                rectangles.Add(layouter.PutNextRectangle(new SizeF(30, 50)));
+            }
+
+            var random = new Random();
+
+            var strs = new List<string>();
+
+            for (var i = 0; i < 100; ++i)
+            {
+                var builder = new StringBuilder();
+                for (var j = 0; j < 5; ++j)
+                    builder.Append((char) random.Next(33, 126));
+                strs.Add(builder.ToString());
+            }
+
+            visualizator.PutWordsInRectangles(strs, rectangles);
+            
+            visualizator.SaveImage();
```
Так, ну здесь тяжело что-то выделить, но смущают следующие вещи: 
1. magic numbers https://en.wikipedia.org/wiki/Magic_number_(programming) 
2. \+ хотелось бы эти числа как-то задавать 
3. Вообще нет говорящих методов, в коде происходит _что-то_

jexinox replied:
Да я даже не думал, что это ревьюить особо будут, поэтому тут просто накидал небольшой генератор случайных строк, 33 - первый номер нормального печатного символа в ascii, а 126 - последний)

neoezop commented on Homework/TagsCloudVisualization/Extensions/PointFExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Drawing;
+using System.Numerics;
```
Лишняя либа (почисти во всём solution-е)

neoezop commented on Homework/TagsCloudVisualization/Extensions/PointFExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+    {
+        public static float DistanceTo(this PointF from, PointF to)
+        {
+            return (float) Math.Sqrt(Math.Pow(Math.Abs(from.X - to.X), 2) + Math.Pow(Math.Abs(from.Y - to.Y), 2));
```
Кажется, что вариант (from.X - to.X) * (from.X - to.X) + (from.Y - to.Y) * (from.Y - to.Y) будет выглядеть лучше (да и вообще Math.Pow используется довольно редко, если в квадрат нужно просто возвести)

neoezop commented on Homework/TagsCloudVisualization/Extensions/RectangleFExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public static class RectangleFExtensions
+    {
+        public static bool IntersectsWithAny(this RectangleF rect, IEnumerable<RectangleF> rectangles)
```
Ну тогда чтобы вообще красиво было можно добавить "Of" в конец

neoezop commented on Homework/TagsCloudVisualization/Layouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,59 @@
+            return rect;
+        }
+
+        public IEnumerable<RectangleF> PutNextRectangles(IEnumerable<SizeF> rectanglesSizes)
```
Давай не будем хранить методы, которые никак не используются (либо навешивать на них комменты с disable-ом инспекции от ide)

neoezop commented on Homework/TagsCloudVisualization/Extensions/RectangleFExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+
+        public static float GetArea(this RectangleF rectangle) => rectangle.Height * rectangle.Width;
+
+        public static PointF GetCenterPoint(this RectangleF rectangle) 
```
Особенно, если это касается вспомогательных методов

neoezop commented on Homework/TagsCloudVisualization/Layouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,59 @@
+    {
+        public readonly HashSet<RectangleF> rectangles;
+        private readonly Point center;
+        private readonly Spiral spiral;
```
Ещё несколько таких замечаний дальше сделаю, но лучше не игнорить сообщения от ide :(
![изображение](https://user-images.githubusercontent.com/49761942/141820304-5783ed5b-669d-4d0e-b5bf-10d1241419d3.png)

neoezop commented on Homework/TagsCloudVisualization/Layouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,59 @@
+                    spiral.CurrentPoint.Y - rectangleSize.Height / 2,
+                    rectangleSize.Height,
+                    rectangleSize.Width);
+            }
```
Чтобы не делать неловких переназначений, можно обернуть в do ... while ...

neoezop commented on Homework/TagsCloudVisualization/Layouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,59 @@
+        
+        public CircularCloudLayouter(Point center)
+        {
+            spiral = new Spiral(center);
```
Сейчас система довольно негибкая, в будущем хотелось бы, чтобы сюда можно было бы пробрасывать и другие алгоритмы расположений. Поэтому, как минимум, тут везде стоит выделить интерфейсы

neoezop commented on Homework/TagsCloudVisualization/Layouters/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,59 @@
+        public IEnumerable<RectangleF> PutNextRectangles(IEnumerable<SizeF> rectanglesSizes)
+            => rectanglesSizes.Select(PutNextRectangle);
+
+        private RectangleF FindRectanglePosition(SizeF rectangleSize)
```
Можно попробовать найти более удачное название

neoezop commented on Homework/TagsCloudVisualization/Layouters/Spiral.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    internal class Spiral
+    {
+        private float radius = 0;
+        private float angle = 0;
```
Ты ведь специально здесь указываешь нули явно? Это хорошо (на мой вкус, явное обозначение таких штук всегда лучше), но ide-шка тут снова сверлит инспекцией 
![изображение](https://user-images.githubusercontent.com/49761942/141821258-7dafacd1-1ee3-41f6-8a56-43cf2a281f68.png)

Никто не защищён от слепого cleanup-а всего solution-а, поэтому такие инспекции лучше всегда выключать
![изображение](https://user-images.githubusercontent.com/49761942/141821378-9af0499a-bfd7-4f14-8edf-17eeb917c4b8.png)

neoezop commented on Homework/TagsCloudVisualization/Layouters/Spiral.cs:
```diff
@@ -0,0 +1,31 @@
+using System.Drawing;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("TagsCloudVisualizationTests")]
```
Круто! Только смотри, такое обычно уносят в папочку Properties в отдельный файл (например, FriendAssemblies), потому что их может быть много и лучше, чтобы они лежали в одном месте

neoezop commented on Homework/TagsCloudVisualization/Visualization/CircularCloudVisualizator.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            path = imageSavingPath;
+
+            bitmap = new Bitmap(800, 600);
```
А почему их нельзя задавать?

neoezop commented on Homework/TagsCloudVisualization/Visualization/CircularCloudVisualizator.cs:
```diff
@@ -0,0 +1,74 @@
+        private Pen pen;
+        private readonly Random random;
+
+        public CircularCloudVisualizator(string imageSavingPath = "./")
```
Давай будем проверять, существует ли такая папка

neoezop commented on Homework/ConsoleAppForUsingTests/Program.cs:
```diff
@@ -0,0 +1,42 @@
+        static void Main(string[] args)
+        {
+            var layouter = new CircularCloudLayouter(new Point(400, 300));
+            var visualizator = new CircularCloudVisualizator();
```
Сейчас картинки по дефолту сохраняются в bin/{config}/net5.0, и за ними вообще не хочется туда лезть... Так что было бы лучше сохранять их в отдельную папку где-то поближе к csproj-у :)

neoezop commented on Homework/TagsCloudVisualization/Visualization/CircularCloudVisualizator.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public string SaveImage()
+        {
+            var randomNumber = random.Next();
```
Интересный способ для нэйминга... Давай избавимся от него (:smiling_imp:) и будем везде передавать имя явно (ну и с дефолтной опцией какой-то, конечно)

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+        public void SetUp()
+        {
+            layouter = new CircularCloudLayouter(center);
+            visualizator = new CircularCloudVisualizator();
```
Для таких картинок стоит завести отдельную папку (почти такой  же коммент, как и выше)

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+            if (TestContext.CurrentContext.Result.Outcome.Status is TestStatus.Failed or TestStatus.Inconclusive)
+            {
+                visualizator.PutRectangles(layouter.rectangles);
+                Console.WriteLine($"Tags cloud visualizaton is saved to {visualizator.SaveImage()}");
```
Давай будем сохранять картинку с именем теста, который упал

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+        [Test]
+        public void PutRectangles_WithoutIntersections()
+        {
+            var rectangles = Generators.RectanglesRandomSizeGenerator()
```
Вообще тут такое дело: рандомайзеры в тестах лучше не использовать, конечно, eсли не хочешь задавить числом. Непонятно, какого результата мы ожидаем, потому что он всегда будет подвержен переменчивости. Давай мы будем чётко указывать проверяемые данные в таких тестах

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        [Test]
+        public void PutNextRectangle_AverageRandomSizeRectanglesCloudDensity_Between65And85Percent()
```
А если и будем тестить что-то со случайными данными, то будем накидвать на такие тесты большое число Repeat-ов через атрибут Repeat в Nunit-е

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+        [Test]
+        public void PutNextRectangle_AverageRandomSizeRectanglesCloudDensity_Between65And85Percent()
+        {
+            var avgDensity = Generators.CenterGenerator()
```
Тут лучше тесты немного переделать. Эти центральные точки хотелось бы видеть в TestCase-ах

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+        public void PutNextRectangle_AverageRandomSizeRectanglesCloudDensity_Between65And85Percent()
+        {
+            var avgDensity = Generators.CenterGenerator()
+                .Select(center => GetCloudDensity(Generators.RectanglesRandomSizeGenerator()))
```
А зачем тут центры, если у нас layouter от обычного центра?

jexinox replied:
Не заметил, переделывал чутка тесты из-за того, что нужно было сделать выкладку, в случае падения и половину перелопатил и не очень хорошо перелопатил

neoezop commented on Homework/TagsCloudVisualizationTests/CircularCloudLayouter_Should.cs:
```diff
@@ -0,0 +1,121 @@
+        [Test]
+        public void PutNextRectangle_AverageSameSizeRectanglesCloudDensity_Between70And90Percents()
+        {
+            var avgDensity = Generators.CenterGenerator()
```
То же, что и сверху

neoezop commented on Homework/TagsCloudVisualizationTests/Generators.cs:
```diff
@@ -0,0 +1,37 @@
+using System.Drawing;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("ConsoleAppForUsingTests")]
```
Ого... Можно убрать, потому что в ConsoleAppForUsingTests ничего из тестов не используется. На будущее: супер странный сценарий. Даже если и надо заиспользовать что-то из тестов, то это хороший повод выделить это в какое-то общее место

neoezop commented on Homework/TagsCloudVisualizationTests/Generators.cs:
```diff
@@ -0,0 +1,37 @@
+
+            for (var i = 0; i < 100; ++i)
+            {
+                // +0.01f is added to avoid NextDouble returned zero
```
Так давай будем использовать фиксированный коэффициент?

jexinox committed:
Fixes in tests and TagsCloudVisualization library

jexinox committed:
Merge remote-tracking branch 'origin/master'

jexinox committed:
CircularCloudVisualizator.cs now saving images to normal name
Tests is saving failed tests to image with failed test name
Little refactor in Console App
